### PR TITLE
Gather information from RestartCount counter

### PR DESCRIPTION
**Pull request by rikatz**
This PR adds a new metric into containers: the restart_count.

This metric can be used in Docker to check how many times a container configured with '--restart=always' has been restarted.

This means some problem like OOM or an application problem (segfault, as an example) and can be used later, with Prometheus to alert about this problem.

This PR uses the Docker API, but can also be extended to other future container engine (or even raw systemd stats, when available).

This is related to issue #1647

### CODE REVIEW

**Comment by dashpole** [comment]
@k8s-bot ok to test

File: container/docker/handler.go [hunk 1]
```diff
@@ -248,6 +251,7 @@ func newDockerContainerHandler(
 	handler.image = ctnr.Config.Image
 	handler.networkMode = ctnr.HostConfig.NetworkMode
 	handler.deviceID = ctnr.GraphDriver.Data["DeviceId"]
+	handler.restartCount = ctnr.RestartCount
```
Comment (dashpole) [inline container/docker/handler.go]: If I read this correctly, you are only updating handler.restartCount when a new docker container handler.  Have you verified that this increases when the container restarts?

File: info/v2/container.go [hunk 2]
```diff
@@ -146,6 +146,8 @@ type ContainerStats struct {
 	Filesystem *FilesystemStats `json:"filesystem,omitempty"`
 	// Task load statistics
 	Load *v1.LoadStats `json:"load_stats,omitempty"`
+	// RestartCount contains how many times a container was restarted automatically
+	RestartCount int `json:"restartcount,omitempty"`
```
Comment (dashpole) [inline info/v2/container.go]: do you convert restartcount from v1 to v2 anywhere?

File: metrics/prometheus.go [hunk 3]
```diff
@@ -135,6 +135,14 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc) *PrometheusCo
 					return metricValues{{value: float64(time.Now().Unix())}}
 				},
 			}, {
+				name:      "container_restart_count",
+				help:      "Counter of times a container has been restarted",
+				valueType: prometheus.CounterValue,
+				getValues: func(s *info.ContainerStats) metricValues {
+					//return metricValues{{value: float64(11111)}}
```
Comment (dashpole) [inline metrics/prometheus.go]: remove

File: container/docker/handler.go (hunk 1 above)
Comment (rikatz) [inline container/docker/handler.go]: @dashpole I've tested this and works fine. The counter is increases each time a container restarts :) This is only a handler of 'how to gather this information', and it's used [here](https://github.com/estaleiro/cadvisor/blob/e42a357488c3b1cb3411b4647d61e73096230bba/container/docker/handler.go#L472)

File: info/v2/container.go (hunk 2 above)
Comment (rikatz) [inline info/v2/container.go]: Is that needed? Didn't know about that. Will check and post a new commit wheter this is necessary.

File: metrics/prometheus.go (hunk 3 above)
Comment (rikatz) [inline metrics/prometheus.go]: The idea here is to make available the times a countainer has been restarted to Prometheus. This is why I've created this metric (and also this PR), so we can check and alert here when a container have a big restart count.

Is this really necessary to be removed?

**Comment by rikatz** [comment]
@dashpole Made the comments for you. Only have to check the v2 API.

About Kubernetes engine, it uses an internal counter and you can export those metrics to prometheus using kube-state-metrics.

About rkt, its API  doesn't support a restart count (as in CoreOS, as an example this is controlled through systemd) and maybe this can be removed.

This is also why this 'metric' is handled only by the Docker Handler, and shows a value of '0' in other systems (like raw and RKT).

Maybe using some conditional in Prometheus metrics would make it less noisy, like if the restart count equals to 0 there is no need to export/show it anyway.

Thanks for your time and your review :)

File: metrics/prometheus.go (hunk 3 above)
Comment (dashpole) [inline metrics/prometheus.go]: just meant the extra comment

File: info/v2/container.go (hunk 2 above)
Comment (dashpole) [inline info/v2/container.go]: [This is where we convert.  Should be simple to implement](https://github.com/google/cadvisor/blob/master/info/v2/conversion.go).

File: container/docker/handler.go (hunk 1 above)
Comment (dashpole) [inline container/docker/handler.go]: Yes, I was more making sure that we create a new dockerContainerHandler each time the container is restarted by docker.  Kubernetes doesn't use docker's restart policies, so I am not familiar with how we handle them.

**Comment by dashpole** [comment]
I am on board with exposing this as a prometheus metric.  I agree this is useful for users monitoring pure docker deployments.  Not exporting if it is 0 would also be nice.

My main concern is that this adds a field to the ContainerStats struct, which is exposed via the api/vX.X/containers endpoint.  For users of other runtimes, or using kubernetes, this additional metric would not only not be useful, but would be misleading, since they would expect this number to equal the kubernetes container restart count.  I would prefer at least renaming this to DockerRestartCount to differentiate it from the kubernetes one.  Generally, we don't explicitly add runtime-specific fields like this to the API.

The ContainerStats struct is currently only being used for general resource usage metrics: CPU, memory, disk, network, and diskio.  I don't think restart count fits into this model.  We don't really have a good way to express a runtime-specific orchestration metric like this.

@timstclair any thoughts on the best way to do this?

**Comment by timstclair** [comment]
I agree with David. I think this metric is more runtime specific, which is something we're trying to move away from. This metric makes more sense if cAdvisor is a general container monitoring solution. It makes less sense if we think of it as a resource monitor. I'm not sure we've concretely defined the scope of cAdvisor anywhere.

**Comment by dashpole** [comment]
@k8s-bot ok to test

**Comment by rikatz** [comment]
@dashpole @timstclair OK, understood :)

I though that creating this metric would make easier to monitor some container problems, but I did understand what's the main focus of cAdvisor now :)

No problems, I can even create a standalone golang program to expose only this, or maybe using metricbeat or zabbix to expose this info also would be fine :)

Anyway, I'm going to make the suggested changes (between today and tomorrow), and if You have any suggestion in how (and if we should) to achieve this in cAdvisor, I'm here available for :)

Thanks for your time guys.

File: container/docker/handler.go (hunk 1 above)
Comment (rikatz) [inline container/docker/handler.go]: Yes, actually the new handler defines how the info is going to be gathered by GetStats function. So I assume this point of review is solved :)

File: metrics/prometheus.go (hunk 3 above)
Comment (rikatz) [inline metrics/prometheus.go]: My bad, removed this :)

**Comment by rikatz** [comment]
@dashpole @timstclair ok, kubernetes exposes this as a label:

``container_cpu_system_seconds_total{[...]container_label_io_kubernetes_container_restartCount="0"[...]} 1.29``

So, if I change the handler from Kubernetes to create this as a label (inside GetStats or something like that), but only for Docker, would this be fine?

The same happens in API v1, but v2 doesn't exposes the Labels, so it have to be checked later as something to be migrated, regardless of being a restart count or not.

This is not the perfect solution for my case (as this is a Label and not a Counter for Prometheus), but as Kubernetes also does this way, I think this is the way to be followed.

May I change some parts of the code to accomplish this way?

Thanks!

**Comment by rikatz** [comment]
@dashpole The changes I've made here insert the restartcount as a label instead of creating a new stats.

This mades easier to implement (also, I don't need to change the API code, and this will be available in APIv2 as someone ports also the labels handler).

Please take a look. I don't think cAdvisor's main feature is to transform some internal Docker Counters into Labels, but I do agree this is probably the best way we can do this instead of changing the stats core.

If, anyway this is not ok, then I'm going to write something else (or maybe fork cAdvisor), as I need this info to Prometheus to check if there is a container in a crash loop or something else (Out of memory, etc), to trigger an alarm to me.

Thanks anyway guys :smile:

File: container/docker/handler.go [hunk 4]
```diff
@@ -31,6 +31,8 @@ import (
 	dockerutil "github.com/google/cadvisor/utils/docker"
 	"github.com/google/cadvisor/zfs"
 
+	"strconv"
```
Comment (dashpole) [inline container/docker/handler.go]: nit: we usually place all of the golang imports (e.g. fmt, time, strconv) together at the top.

**Comment by dashpole** [comment]
@timstclair this looks acceptable to me.  It would never show up in kubernetes, since we don't use docker's restart feature.
WDYT?

File: container/docker/handler.go (hunk 4 above)
Comment (rikatz) [inline container/docker/handler.go]: Yeah, this is vscode fault. Corrected this :)

**Comment by rikatz** [comment]
@dashpole Anything about this? Thanks :D

**Comment by rikatz** [comment]
@dashpole friendly ping :)

**Comment by dashpole** [comment]
Yeah, sorry.  I plan to merge this once we get our test-infra stuff figured out...  Just hang tight.

**Comment by dashpole** [comment]
would you mind squashing?

**Comment by rikatz** [comment]
@dashpole Sure, will do this.

**Comment by rikatz** [comment]
Did as asked :) Anything else you need I'm here available to help! And thanks for your time :D

**Comment by rikatz** [comment]
Thanks :)

**Pull request merged**